snorkel/analysis/scorer.py

```diff
@@ -98,3 +99,54 @@ def score(
                 metric_dict[metric_name] = score
 
         return metric_dict
+
+    def score_slices(
+        self,
+        S: np.recarray,
+        golds: np.ndarray,
+        preds: np.ndarray,
+        probs: np.ndarray,
+        as_dataframe: bool = False,
+    ) -> Union[Dict[str, Dict[str, float]], pd.DataFrame]:
+        """Calculate user-specified and/or user-defined metrics overall + slices.
+
+        Parameters
+        ----------
+        S
+            A recarray with entries of length n_examples corresponding to slice names
+        golds
+            Gold (aka ground truth) labels (integers)
+        preds
+            Predictions (integers)
+        probs:
+            Probabilities (floats)
+        as_dataframe
+            A boolean indicating whether to return results as pandas ``DataFrame`` (True)
+            or dict (False)
+
+        Returns
+        -------
+        Union[Dict, pd.DataFrame]
+            A dictionary mapping slice_name to metric names to metric scores
+            or aforementioned dictionary formatted as pandas ``DataFrame``
+        """
+        assert S.shape[0] == len(golds) == len(preds) == len(probs)
+        if not isinstance(S, np.recarray):
+            raise ValueError(f"S should be {np.recarray} type.")
+
+        # Include overall metrics
+        metrics_dict = dict()
+        metrics_dict.update({"overall": self.score(golds, preds, probs)})
+
+        # Include slice metrics
+        slice_names = S.dtype.names
+        for slice_name in slice_names:
+            mask = S[slice_name].astype(bool)
+            metrics_dict.update(
```

Comment:
why update? the entries of slice_names should be unique

Reply:
as in, do `metrics_dict[slice_name] = self.score(...)`

Reply:
👍